stream_pooling.go

```diff
@@ -0,0 +1,181 @@
+	pb "github.com/libp2p/go-libp2p-kad-dht/pb"
+	inet "github.com/libp2p/go-libp2p-net"
+	peer "github.com/libp2p/go-libp2p-peer"
+	"golang.org/x/xerrors"
+)
+
+func (dht *IpfsDHT) getPoolStream(ctx context.Context, p peer.ID) (_ *poolStream, reused bool, _ error) {
+	dht.streamPoolMu.Lock()
+	for ps := range dht.streamPool[p] {
+		dht.deletePoolStreamLocked(ps, p)
+		if ps.err() != nil {
+			// Stream went bad and hasn't deleted itself yet.
+			continue
+		}
+		dht.streamPoolMu.Unlock()
+		return ps, true, nil
+	}
+	dht.streamPoolMu.Unlock()
+	ps, err := dht.newPoolStream(ctx, p)
+	return ps, false, err
+}
+
+func (dht *IpfsDHT) putPoolStream(ps *poolStream, p peer.ID) {
+	dht.streamPoolMu.Lock()
+	defer dht.streamPoolMu.Unlock()
+	if ps.err() != nil {
+		return
+	}
+	if dht.streamPool[p] == nil {
+		dht.streamPool[p] = make(map[*poolStream]struct{})
+	}
+	dht.streamPool[p][ps] = struct{}{}
+}
+
+func (dht *IpfsDHT) deletePoolStream(ps *poolStream, p peer.ID) {
+	dht.streamPoolMu.Lock()
+	dht.deletePoolStreamLocked(ps, p)
+	dht.streamPoolMu.Unlock()
+}
+
+func (dht *IpfsDHT) deletePoolStreamLocked(ps *poolStream, p peer.ID) {
+	delete(dht.streamPool[p], ps)
+	if len(dht.streamPool) == 0 {
+		delete(dht.streamPool, p)
+	}
+}
+
+func (dht *IpfsDHT) newPoolStream(ctx context.Context, p peer.ID) (*poolStream, error) {
+	s, err := dht.newStream(ctx, p)
+	if err != nil {
+		return nil, xerrors.Errorf("opening stream: %w", err)
+	}
+	ps := &poolStream{
+		stream: s,
+		w:      newBufferedDelimitedWriter(s),
+		r:      pbio.NewDelimitedReader(s, inet.MessageSizeMax),
+		m:      make(chan chan *pb.Message, 1),
+	}
+	ps.onReaderErr = func() {
+		ps.reset()
+		dht.deletePoolStream(ps, p)
+	}
+	ps.onReady = func() {
+		dht.putPoolStream(ps, p)
+	}
+	go ps.reader()
+	return ps, nil
+}
+
+type poolStream struct {
+	stream interface {
+		Reset() error
+	}
+	w bufferedWriteCloser
+	r pbio.ReadCloser
+	// Called when the reader dies.
+	onReaderErr func()
+	// Called after reading an expected message.
+	onReady func()
+
+	// Synchronizes m and readerErr.
+	mu sync.Mutex
+	// Receives channels to send responses on.
+	m         chan chan *pb.Message
+	readerErr error
+}
+
+func (me *poolStream) reset() {
+	me.stream.Reset()
+}
+
+func (me *poolStream) send(m *pb.Message) (err error) {
+	defer func() {
+		if err != nil {
+			log.Printf("error sending message: %v", err)
+		}
+	}()
+	if err := me.w.WriteMsg(m); err != nil {
+		return xerrors.Errorf("writing message: %w", err)
+	}
+	if err := me.w.Flush(); err != nil {
+		return xerrors.Errorf("flushing: %w", err)
+	}
+	return nil
+}
+
+func (me *poolStream) request(ctx context.Context, req *pb.Message) (*pb.Message, error) {
+	replyChan := make(chan *pb.Message, 1)
+	me.mu.Lock()
+	if err := me.err(); err != nil {
+		me.mu.Unlock()
+		return nil, err
+	}
+	select {
+	case me.m <- replyChan:
+	default:
+		me.mu.Unlock()
+		panic("message pipeline full")
+	}
+	me.mu.Unlock()
+	err := me.send(req)
+	if err != nil {
```

Comment:
We're leaving the `replyChan` in `me.m` in this case. A subsequent call to `request` before the read loop exits will panic.

Reply:
Yes, but we don't put the poolStream back in the pool if we get an error back from request. This is by design, if we did, we would be pipelining.

Reply:
Ah, got it.